### PR TITLE
Defer loading additional properties

### PR DESCRIPTION
**Pull request by yongyan-gh**
The properties of SARIF results: Locations/RelatedLocations/Stacks/CodeFlows(AnalysisSteps)/PropertyBags are not needed when loading SARIF results into Error List. They can be populated when the data are needed, like before SARIF explorer display result details.

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs [hunk 1]
```diff
@@ -391,7 +340,11 @@ public string SelectedTab
         public ObservableCollection<KeyValuePair<string, string>> Properties { get; }
 
         [Browsable(false)]
-        public bool HasDetails => this.Locations.Any() || this.RelatedLocations.Any() || this.AnalysisSteps.Any() || this.Stacks.Any() || this.Fixes.Any();
+        public bool HasDetails => this.SarifResult.Locations?.Any() == true ||
+                                  this.SarifResult.RelatedLocations?.Any() == true ||
+                                  this.SarifResult.CodeFlows?.Any() == true ||
+                                  this.SarifResult.Stacks?.Any() == true ||
+                                  this.SarifResult.Fixes?.Any() == true;
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:347]: Order by length

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs [hunk 2]
```diff
@@ -39,7 +39,7 @@ public static LocationCollection ToThreadFlowLocationCollection(this CodeFlow co
 
         public static AnalysisStep ToAnalysisStep(this CodeFlow codeFlow, Run run, int resultId, int runIndex)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
+            // ThreadHelper.ThrowIfNotOnUIThread();
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs:42]: If not useful, remove it. Also remove the blank line.

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/StackFrame.Extensions.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@ public static StackFrameModel ToStackFrameModel(this StackFrame stackFrame, int
         {
             var model = new StackFrameModel(resultId, runIndex);
 
-            model.FullyQualifiedLogicalName = stackFrame.Location.LogicalLocation.FullyQualifiedName;
+            model.FullyQualifiedLogicalName = stackFrame.Location?.LogicalLocation?.FullyQualifiedName;
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Sarif/StackFrame.Extensions.cs:15]: Do we have a test covering this nullable parts? If not, pls, add.

**Review by eddynaka — CHANGES_REQUESTED** [review]

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/StackFrame.Extensions.cs (hunk 3 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Sarif/StackFrame.Extensions.cs:15]: found the issue by new added SarifErrorListItemTests, also added unit tests for StackFrameExtensions.ToStackFrameModel itself.

**Review by eddynaka — APPROVED** [review]
:shipit:

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs]: Why was this removed exactly?

File: src/Sarif.Viewer.VisualStudio.UnitTests/Sarif/StackFrameExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            var stackFrameModel = stackFrame.ToStackFrameModel(resultId, runIndex);
+            stackFrameModel.Should().NotBeNull();
+            stackFrameModel.ResultId.Should().Be(resultId);
+            stackFrameModel.RunIndex.Should().Be(runIndex);
+            stackFrameModel.Module.Should().BeNull();
+            stackFrameModel.Message.Should().BeEquivalentTo("stack frame location 1");
+            stackFrameModel.FullyQualifiedLogicalName.Should().BeEquivalentTo(@"\root\FQDN\path-to\file.ext");
+            stackFrameModel.FilePath.Should().BeEquivalentTo("path/to/file.cpp");
+            stackFrameModel.Region.Should().NotBeNull();
+            stackFrameModel.Region.StartLine.Should().Be(10);
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio.UnitTests/Sarif/StackFrameExtensionsTests.cs:90]: >StartLine [](http://example.com/codeflow?start=35&length=9)

This test code is not very maintainable, as it creates multiple copies of literals. You could have cleaned this up to some degree by consulting values from the stackFrame object itself, which you retained. 

stackFrameModel.Region.Startlein.Should().Be(stackFrame.Location.PhysicalLocation.Region.StartLine).

In general, when you find yourself repeating hard-coded strings and values, like your relative path @"\root\", etc., that's a clue to try to reduce the code to a single copy, i.e., authoring the code where the literal value only appears once.

This is a basic quality standard that we enforce through-out our codebase (and which is very common in many projects).

Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.UnitTests/Sarif/StackFrameExtensionsTests.cs:90]: thanks for the feedback, refactor the test case and move the verification into a common function which does not depends on hardcoded values. please review pr: [https://github.com/microsoft/sarif-visualstudio-extension/pull/448](https://github.com/microsoft/sarif-visualstudio-extension/pull/448)

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs (hunk 2 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Sarif/CodeFlow.Extensions.cs]: this function does not invoke anything related to UI elements, so its not required.

**Pull request merged**